component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -0,0 +1,202 @@
+                        "attributes. Hence returning false.");
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Retrieves the client ID which is extracted from incoming request.
+     *
+     * @param request                 HttpServletRequest.
+     * @param bodyParams              Body parameter map of the incoming request.
+     * @param oAuthClientAuthnContext OAuthClientAuthentication context.
+     * @return Client ID of the OAuth2 client.
+     * @throws OAuthClientAuthnException OAuth client authentication Exception.
+     */
+    @Override
+    public String getClientId(HttpServletRequest request, Map<String, List> bodyParams, OAuthClientAuthnContext
+            oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        Map<String, String> stringContent = getBodyParameters(bodyParams);
+        oAuthClientAuthnContext.setClientId(stringContent.get(OAuth.OAUTH_CLIENT_ID));
+        return oAuthClientAuthnContext.getClientId();
+    }
+
+    private boolean clientIdExistsAsParam(Map<String, List> contentParam) {
+
+        Map<String, String> stringContent = getBodyParameters(contentParam);
+        return (StringUtils.isNotEmpty(stringContent.get(OAuth.OAUTH_CLIENT_ID)));
+    }
+
+    /**
+     * Check for the existence of a valid certificate in required format in the request attribute map.
+     *
+     * @param request HttpServletRequest which is the incoming request.
+     * @return Whether a certificate exists or not.
+     */
+    private boolean validCertExistsAsAttribute(HttpServletRequest request) {
+
+        Object certObject = request.getAttribute(JAVAX_SERVLET_REQUEST_CERTIFICATE);
+        return (certObject instanceof X509Certificate[] || certObject instanceof X509Certificate);
+    }
+
+    /**
+     * Authenticate the client by comparing the public key of the registered public certificate against the public
+     * key of the certificate presented at TLS hand shake for authentication.
+     *
+     * @param registeredCert X.509 certificate registered at service provider configuration.
+     * @param requestCert    X.509 certificate presented to server during TLS hand shake.
+     * @return Whether the client was successfully authenticated or not.
+     */
+    protected boolean authenticate(X509Certificate registeredCert, X509Certificate requestCert)
+            throws OAuthClientAuthnException {
+
+        boolean trustedCert = false;
+        try {
+            String publicKeyOfRegisteredCert = MutualTLSUtil.getThumbPrint(registeredCert);
+            String publicKeyOfRequestCert = MutualTLSUtil.getThumbPrint(requestCert);
+            if (StringUtils.equals(publicKeyOfRegisteredCert, publicKeyOfRequestCert)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Client certificate thumbprint matched with the registered certificate thumbprint.");
+                }
+                trustedCert = true;
```

Comment:
There is no log if this mach doesn't pass

Reply:
Fixed